docs/resources/server/recipes/custom.md

```diff
@@ -0,0 +1,134 @@
+
+To be able to obtain a recipe instance from the manager, `#matches` must return true. This method checks against the provided container to see whether the associated inputs are valid. `Ingredient`s can be used for validation by calling `Ingredient#test`.
+
+If the recipe has been chosen, it is then built using `#assemble` which may use data from the inputs to create the result.
+
+!!! tip
+    `#assemble` should always produce a unique `ItemStack`. If unsure whether `#assemble` does so, call `ItemStack#copy` on the result before returning.
+
+Most of the other methods are purely for integration with the recipe book.
+
+```java
+public record ExampleRecipe(Ingredient input, int data, ItemStack output) implements Recipe<Container> {
+  // Implement methods here
+}
+```
+
+RecipeType
+----------
+
+`RecipeType` is responsible for defining the category or context the recipe will be used within. For example, if a recipe was going to be smelted in a furnace, it would have a type of `RecipeType#SMELTING`.  Being blasted in a blast furnace would have a type of `RecipeType#BLASTING`.
+
+If none of the existing types match what context the recipe will be used within, then a new `RecipeType` must be [registered][nonforge].
+
+The `RecipeType` instance must then be returned by `Recipe#getType` in the new recipe subtype.
+
+```java
+// For some RecipeType EXAMPLE_TYPE
+// In ExampleRecipe
+@Override
+public RecipeType<?> getType() {
+  return EXAMPLE_TYPE;
+}
+```
+
+RecipeSerializer
+----------------
+
+A `RecipeSerializer` is responsible for decoding JSONs and communicating across the network for an associated `Recipe` subtype. Each recipe decoded by the serializer is saved as a unique instance within the `RecipeManager`.
+
+Only three methods need to be implemented for a `RecipeSerializer`:
+
+Method | Description
+:---: | :---
+fromJson | Decodes a JSON into the `Recipe` subtype.
+toNetwork | Encodes a `Recipe` to the buffer to send to the client. The recipe identifier does not need to be encoded.
+fromNetwork | Decodes a `Recipe` from the buffer sent from the server. The recipe identifier does not need to be decoded.
+
+A `RecipeSerializer` must be [registered][forge].
+
+!!! tip
+    For ease of convenience, the `RecipeSerializer` subtype can extend `ForgeRegistryEntry` to implement the methods within `IForgeRegistryEntry`.
+
+```java
+public class ExampleSerializer extends ForgeRegistryEntry<RecipeSerializer<?>> implements RecipeSerializer<ExampleRecipe> {
+  // Implement methods here
+}
+```
+
+!!! tip
+    There are some useful methods to make reading and writing data for recipes easier. `Ingredient`s can use `#fromJson`, `#toNetwork`, and `#fromNetwork` while `ItemStack`s can use `CraftingHelper#getItemStack`, `FriendlyByteBuf#writeItem`, and `FriendlyByteBuf# readItem`.
```

Comment:
For the sake of spacing out admonitions, please move this to the end of the section.